### ISSUE TITLE
rclone with backblaze b2 remote and --files-from option seems to ignore --fast-list option.

### ISSUE THREAD

**Issue by usbpc** [ncw/rclone#3095]
#### What is the problem you are having with rclone?
I'm trying to sync a lot of small files (~1,200,00 files with just ~45GB in total size) from my local machine to backblaze b2. Since I know exactly what files have changed I have a text file with all the changed filenames and am using the `--files-from` to just sync those files and speed up the process a lot. Since I also want to minimize b2 api calls I also specify `--fast-list` that should reduce api calls to backblaze (according to rclone documentation) and that worked untill I tried to combine it with `--fast-list` not it seems very slow and used up way more api calls then before, right up to my daily cap in fact.

#### What is your rclone version (output from `rclone version`)
rclone v1.46
- os/arch: linux/amd64
- go version: go1.11.5

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Ubuntu bash in WSL Windows 10 64bit

####  Which cloud storage system are you using? (eg Google Drive)
Backblaze b2

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
`rclone --fast-list --files-from ./mapChanges.txt --b2-hard-delete --transfers 32 sync /mnt/e/servermap_optimizedImages b2:usbCraft/map/ -P`

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
Unfortunatly I don't have this right now and also could not test it with the latest beta yet since I've used up my Daily Class C transactions on backblaze and don't want to increase them today, I will however run the same command with debug output and in the latest beta tomorrow after the cap resets.

**Comment by usbpc** [comment]
As promised output with beta and `-vv`

```
$ rclone version
rclone v1.46.0-112-g1c301f9f-beta
- os/arch: linux/amd64
- go version: go1.12.1
```
`rclone --fast-list --files-from ./mapChanges.txt --b2-hard-delete --transfers 32 sync /mnt/e/servermap_optimizedImages b2:usbCraft/map/ -vv`

```
2019/04/06 02:29:42 DEBUG : rclone: Version "v1.46.0-112-g1c301f9f-beta" starting with parameters ["rclone" "--fast-list" "--files-from" "./mapChanges.txt" "--b2-hard-delete" "--transfers" "32" "sync" "/mnt/e/servermap_optimizedImages" "b2:usbCraft/map/" "-vv"]
2019/04/06 02:29:42 DEBUG : Using config file from "/home/kjh/.config/rclone/rclone.conf"

2019/04/06 02:30:44 INFO  :
Transferred:             0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:                 0 / 0, -
Transferred:            0 / 0, -
Elapsed time:     1m15.3s

2019/04/06 02:31:44 INFO  :
Transferred:             0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:                 0 / 0, -
Transferred:            0 / 0, -
Elapsed time:     2m15.3s

2019/04/06 02:32:44 INFO  :
Transferred:             0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:                 0 / 0, -
Transferred:            0 / 0, -
Elapsed time:     3m15.3s

2019/04/06 02:33:44 INFO  :
Transferred:             0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:                 0 / 0, -
Transferred:            0 / 0, -
Elapsed time:     4m15.3s```

In that time it used about 11,000 Class C b2 transactions.

**Comment by usbpc** [comment]
Interestingly it works as expected in an older version:

```
$ rclone -V
rclone v1.43
- os/arch: linux/amd64
- go version: go1.11
```

It dosen't work quite as fast as I'd like but it dosen't use a way to big amount of b2 class C transactions.

What I think is making it slower has something to do with  #2639 as the local file system dosen't benefit from `--fast-list` but rclone still does a recursive listing of all files even tho `--files-from` is specified.

But it does use fewer transactions and with `--files-from` it is way faster cause it dosen't really compare all files it just compares everything against the `--files-from` whitelist.

**Comment by ncw** [comment]
This change was made in c5ac96e9e7afd2e11b69233ed3671e8ad05d24a9 which was released in v1.45.

> Make --files-from only read the objects specified and don't scan directories
> 
> Before this change using --files-from would scan all the directories
> that the files could possibly be in causing rclone to do more work
> that was necessary.
> 
> After this change, rclone constructs an in memory tree using the
> --fast-list mechanism but from all of the files in the --files-from
> list and without scanning any directories.
> 
> Any objects that are not found in the --files-from list are ignored
> silently.
> 
> This mechanism is used for sync/copy/move (march) and all of the
> listing commands ls/lsf/md5sum/etc (walk).

There is a tradeoff for doing `--files-from` like this and just doing a full file system traverse (eg with `--fast-list`).  It should be faster but I suspect for big lists it won't be faster.

How many files do you have in your list?

This problem came up on the forum too (with google drive) and there you'll find a beta with a flag `--files-from-traverse` to re-enable the old behavior.

https://forum.rclone.org/t/using-files-from-with-drive-hammers-the-api/8726

https://beta.rclone.org/branch/v1.46.0-019-gf97cbd3b-fix-files-from-traverse-beta/

I wonder whether I made a mistake with the new `--files-from` behaviour, or should add some heuristic to turn it on and off...

Perhaps `--fast-list` should turn it off and force it to use the whole listing... That would fix this problem and the one in the forum.

**Comment by ncw** [comment]
I warmed up the beta here

https://beta.rclone.org/branch/v1.46.0-114-geb6ff029-fix-files-from-traverse-beta/ (uploaded in 15-30 mins)

Enable the workaround with `--files-from-traverse`

**Comment by usbpc** [comment]
Nice, good to know when it was changed. Looking through the backblaze API it seems like there would be no good way for rclone to know if listing all files or just the files specified with `--files-from` would be cheaper.

The way to do this would be to see how many files are in the bucket in total and that information dosen't seem to be available in the b2 api.

With that in mind I feel like the best way to handle the situation is kind of what you are doing with the beta with allowing the use to choose how to handle the situation since they can easily decide what would be better for them.

Also is the way I'm doing things actually a good way to do it? What I'm trying to do is update a rendered map of my Minecraft server this way, that means that some files will have changed each time and there may also be files that are not already in the b2 bucket. And in the bucket currently there are 1.2 Million files and about 200_000 files are changed/created every time I update the map.

**Comment by ncw** [comment]
> With that in mind I feel like the best way to handle the situation is kind of what you are doing with the beta with allowing the use to choose how to handle the situation since they can easily decide what would be better for them.

I'm leaning towards making the previous behavior the default and making the new behaviour selectable with the `--no-traverse` flag which has exactly the correct semantics.

> Also is the way I'm doing things actually a good way to do it? What I'm trying to do is update a rendered map of my Minecraft server this way, that means that some files will have changed each time and there may also be files that are not already in the b2 bucket. And in the bucket currently there are 1.2 Million files and about 200_000 files are changed/created every time I update the map.

If you are updating such a large % of files (16%) you might just be better doing a `sync` with `--fast-list` and forgetting the `--files-from` list.  It probably depends on whether the updates are spread all over, or whether they are concentrated into a subset of directories.

**Comment by ncw** [comment]
I've fixed this here - testing appreciated :-) I changed tack and made the default behavior as it was, and you'll need the `--no-traverse flag` to enable the new behaviour.  So in your case you don't want the `--no-traverse` flag.
    filter: Make --files-from traverse as before unless --no-traverse is set
    
    In c5ac96e9e7 we made --files-from only read the objects specified and
    don't scan directories.
    
    This caused problems with Google drive (very very slow) and B2
    (excessive API consumption) so it was decided to make the old
    behaviour (traversing the directories) the default with --files-from
    and use the existing --no-traverse flag (which has exactly the right
    semantics) to enable the new non scanning behaviour.


https://beta.rclone.org/branch/v1.46.0-114-gc7bf473c-fix-files-from-traverse-beta/ (uploaded in 15-30 mins)

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.47
